### ISSUE TITLE
Support bash style alias's to support tab completion in command string aliases

### ISSUE THREAD

**Issue by TylerLeonhardt** [PowerShell/PowerShell#12962]
## Summary of the new feature/enhancement

In bash, you can do alias's like so:
```bash
alias gb="git branch"
```
which is really useful for typing less.

In PowerShell we can do:
```pwsh
function gb {
    git branch @args
}
```

But here's why bash's are so much better:
![image](https://user-images.githubusercontent.com/2644648/84602391-e49cb480-ae3b-11ea-8b6e-8713863af04b.png)

tab completion _still works_. Where as for PowerShell they don't.

By not supporting this in some way, it means that alias's to native executables are better in bash because the experience is more fluid.

Also, the function approach that we have means we lose out on the work @rjmholt did in UnixCompleters.

## Proposed technical implementation details (optional)

Perhaps a new kind of alias that will funnel tab completion all the way down to the alias'd command string?

```pwsh
Set-Alias gb 'git branch' -AliasType CommandString # vs ... CommandLiteral or something... naming is hard but you get the idea
```

**Comment by p0W3RH311** [comment]
```
In PowerShell we can do:

function gb {
    git branch @args
}

```
and if  someone add   **gb** alias  it  shadow  your **gb** function.  dont work as expected.

**Comment by SeeminglyScience** [comment]
This is pretty much the first thing people (including myself) try to do with aliases. I've seen countless threads/discord conversations with a ton of folks tripping on this.

I'm tempted to say that `Set-Alias` should do this automatically if the definition:

1. Has spaces
1. Doesn't resolve to a existing command
1. Has a first "part" that *does* resolve

**Comment by doctordns** [comment]
> This is pretty much the first thing people (including myself) try to do with aliases. I've seen countless threads/discord conversations with a ton of folks tripping on this.

Really? The first thing you do is to do complex aliasing? 

An alias is a short cut to a command name only and has been since V1. And in all the years I have moderated the PowerShell forum on Spiceworks this is not a highly requested feature. There have been requests but certainly not many.

I'd say this might be a nice to have - but only if the implementation does not break things.

**Comment by rjmholt** [comment]
If we pull on this string a bit:

- How does the AST capture something like `set-alias gb 'git branch'`, what gets passed in to existing mechanisms for argument completion or pseudoparameterbinding?
- What happens in the following cases:
  - `set-alias gb 'git bran'`
  - `set-alias gc 'git commit --a'`
  - `set-alias gc 'git commit --'`
  - `set-alias gi 'get-item -path'`
  - `set-alias gi 'get-item -f'`
  - `set-alias gi 'get-item -'`

Basically, I'd like to understand where the breakdown lies between PowerShell's parser/AST design vs bash's expand and execute approach. Once we have a better understanding of that, we'll be better placed to evaluate whether this feature is achievable.

**Comment by SeeminglyScience** [comment]
@rjmholt I don't think the parser would need to change.  Just the parameter binder and possibly a new `CommandInfo` (or alternatively, just some different code in `AliasInfo.Parameters`).  Well I say "just" but it's a pretty substantial work item, though I think it's worth it.

For native command argument completers, it might be fine to just expand the text and reparse before invoking them.  Kinda dirty, but maybe doable just for completions.

**Comment by TylerLeonhardt** [comment]
@doctordns 

> Really? The first thing you do is to do complex aliasing?

Yes. People who came from bash/zsh/fish will do this.

Take a look at this guide:
https://www.netlify.com/blog/2020/04/12/speed-up-productivity-with-terminal-aliases/

every single example is complex in the sense that it's a command string and not a particular command.

**Comment by SeeminglyScience** [comment]
> Really? The first thing you do is to do complex aliasing?

I'm pretty sure the first alias I tried to make way back when was to `ipconfig /all`.  I know that fits well into a function as well, but I don't think I'd call `Set-Alias dap 'dotnet add package'` all that complex.

> And in all the years I have moderated the PowerShell forum on Spiceworks this is not a highly requested feature. There have been requests but certainly not many.

Haven't heard of Spiceworks, but I see it a lot on reddit/discord.  It's up there with `Select-Object -Property` vs `ExpandProperty` in starter walls.

> I'd say this might be a nice to have - but only if the implementation does not break things.

That's fair, luckily I'm pretty sure breaks won't be needed for it.

**Comment by rkeithhill** [comment]
Perhaps a more PowerShelly way to implement this would be to add a parameter like `-Parameters` e.g:
```powershell
set-alias gc git -parameters commit, "--a"
# OR maybe parameter allows a single string with all parameters?
set-alias gc git -parameters "commit --a"
```

**Comment by doctordns** [comment]
That would work for many cases, but doesn't handle 'weird' syntax, so you probably still need to have --% for those any way. I like the & --% idea as it reuses the existing operator.

**Comment by vexx32** [comment]
It would handle it just the same way as you'd need to handle it even if there wasn't a separate `-Parameters` thing to work with.

You'd probably have to wrap everything in quotes there; you can't directly use `%--` for this since you're passing data to a cmdlet, not a native executable, so you don't really have the option of not using PS's parsing rules for the input.

**Comment by r-darwish** [comment]
Another approach is Fish [abbreviations](https://fishshell.com/docs/current/cmds/abbr.html). Implementing them might be simpler in terms of auto completion because they work by expanding the alias at the prompt itself.

**Comment by daxian-dbw** [comment]
Quote from https://github.com/PowerShell/PSReadLine/issues/1753#issuecomment-703863585:
> IMO this is more of an editor feature (command line or other) that shouldn't require any PowerShell engine support.

We may consider supporting this in PSReadLine instead of PowerShell engine, so that it stays as an interactive feature only and won't be supported in a script execution.

**Comment by TylerLeonhardt** [comment]
@daxian-dbw this means that intellisense in vscode won't get this feature.

**Comment by rkeithhill** [comment]
@TylerLeonhardt True, but then PSSA will just turn around and tell you that you shouldn't be using aliases in scripts.  😉

**Comment by ratijas** [comment]
> > Really? The first thing you do is to do complex aliasing?
> 
> Yes. People who came from bash/zsh/fish will do this.

I'd like to add, this difference in mindsets comes from the fact that commands in UNIX land (unlike PowerShell) are usually already short enough that you won't need to alias them.  So the primary use case indeed is for stacking up some commonly used arguments.

@doctordns

**Comment by liang2zhu1** [comment]
Just want to add a pressing scenario that can't easily implement without aliasing - 

Ttrying to implement a 'Take top one' function and following would only work functionally but have bad perf. 
```
PS C:\Users\liangzhu> function toponef {$input | Select-Object -First 1}
PS C:\Users\liangzhu> 1,2,3 | toponef
1
```
This brings a performance penalty in the case of pipeline input - if I run `dir c:\ -s | topeonef` it will get to wait for the base stream completes (Compare to that - `dir c:\ -s | Selelct-Object -First 1` works just fine doing early termination) 

More in #15855 on the technical challenge to get parity in function

**Comment by iSazonov** [comment]
> Just want to add a pressing scenario that can't easily implement without aliasing -

"filter" function works more fast. :-)

**Comment by kilasuit** [comment]
reopening as I think there is value in having this in future irrespective of however it gets implemented

**Comment by mklement0** [comment]
I haven't followed this thread in detail, so there may be a better solution, but note that you can pass a `[byte[]]` array directly to the `-Body` argument, so to emulate the 7.3- behavior you can try to ISO-8859-1-encode the entire `$body` string:

```powershell
$body = [System.Text.Encoding]::Latin1.GetBytes((
    "--$boundary",
    "Content-Disposition: form-data; name=`"PWD`"`n",
    $certPw,
    "--$boundary",
    "Content-Disposition: form-data; name=`"FILE`"; filename=`"certificate.pfx`"",
    "Content-Type: application/x-pkcs12`n",
    [System.Text.Encoding]::Latin1.GetString($bytes),
    "--$boundary--`n"
  ) -join "`n")
```

Note that any single-byte, 8-bit, fixed-width text encoding (all of whose code points must be valid) can be used to losslessly encode arbitrary binary data.

ISO-8859-1 encoding (which is closely related to, but not the same as Windows-1252) is special in that it is a true subset of Unicode, covering the 8-bit range of Unicode code points (`` (NULL, [`U+0000`](http://www.fileformat.info/info/unicode/char/0)) through `ÿ` (LATIN SMALL LETTER Y WITH DIAERESIS, [`U+00FF`](http://www.fileformat.info/info/unicode/char/ff))

As such, you can cast arbitrary binary data to a .NET string, and the Unicode code points of the resulting characters will reflect the original byte values; e.g.:

```powershell
[byte[]] $bytes = 0..255

$str = [string]::new([char[]] $bytes)

# -> 0..255
[int[]] [char[]] $str
```

**Comment by mklement0** [comment]
Let me try a summary of sorts:

Supporting Bash-style alias definitions would provide the following benefits:

* Ease of definition, closer alignment with POSIX-compatible shells.
  * The most frictionless syntax is the one suggested by @SeeminglyScience [above](https://github.com/PowerShell/PowerShell/issues/12962#issuecomment-644740948), so that, say 
 `Set-alias gb 'git branch'` would just work, with no need for additional parameters.

* If implemented that way, implicit _pipeline-input_ support with support for _streaming_ processing (line by line), whereas using a `function` to support that currently requires a cumbersome implementation as a _proxy function_  based on a steppable pipeline (because a function based on `$input` invariably collects all input up front; note that use of a `filter` is _not_ an option in this case).

The tab-completion aspect is really an independent one, and for current aliases, which are merely name mappings, already works for _PowerShell_ target commands, but not for _external (native) programs_.

  * For instance, say you've used `Register-ArgumentCompleter` to define completions for `pwsh`, and you've also defined `Set-Alias psh pwsh`; currently, alias `psh` then unexpectedly does _not_ use the tab-completions defined for `pwsh`.
  
  * This asymmetry should be fixed: #21609

Given that the proposal at hand also requires the enhanced aliases to know the external executable name, fixing the above asymmetry could also provide tab-completion support for enhanced aliases.

**Comment by iRon7** [comment]
Another benefit would be that the PSScriptAnalyzer is able to identify and (automatically) replace aliases (in an IDE as VSCode) in formal scripts.

Besides, a possible other way to implement such a feature could be by adding a  `-Definition` parameter that is mutual exclusive with the `-Value` parameter, also resides at position `1`, and only accepts ScriptBlocks. Which would than support syntaxes as:

```PowerShell
Set-Alias gc { git -parameters commit, "--a" }
```

(I guess this is also safer which regards to code injections)
